### PR TITLE
ui/map: cache navigation images at startup

### PR DESCRIPTION
**Pull request by deanlee**
Cache images to avoid reloading, scaling, and flipping (rhd) images with each refresh.it can also prevents Qlable from redrawing the pixmap when its content has not changed.

### CODE REVIEW

File: selfdrive/ui/qt/maps/map.cc [hunk 1]
```diff
@@ -484,27 +506,14 @@ void MapInstructions::updateInstructions(cereal::NavInstruction::Reader instruct
...
-        fn.replace("right", "left");
-      }
-    }
-
-    QPixmap pix(fn);
-    if (is_rhd) {
-      pix = pix.transformed(QTransform().scale(-1, 1));
-    }
-    icon_01->setPixmap(pix.scaledToWidth(200, Qt::SmoothTransformation));
+    bool rhd  = is_rhd && fn.contains("_left") || fn.contains("_right");
```
Comment (incognitojam) [inline selfdrive/ui/qt/maps/map.cc:515]:
```suggestion
    bool rhd = is_rhd && (fn.contains("_left") || fn.contains("_right"));
```

Comment (incognitojam) [inline selfdrive/ui/qt/maps/map.cc:515]:
```suggestion
    bool rhd = is_rhd && fn.contains("_left") || fn.contains("_right");
```

File: selfdrive/ui/qt/maps/map.cc [hunk 2]
```diff
@@ -460,27 +482,14 @@ void MapInstructions::updateInstructions(cereal::NavInstruction::Reader instruct
...
-        fn.replace("right", "left");
-      }
-    }
-
-    QPixmap pix(fn);
-    if (is_rhd) {
-      pix = pix.transformed(QTransform().scale(-1, 1));
-    }
-    icon_01->setPixmap(pix.scaledToWidth(200, Qt::SmoothTransformation));
+    bool rhd = is_rhd && fn.contains("_left") || fn.contains("_right");
```
Comment (adeebshihadeh) [inline selfdrive/ui/qt/maps/map.cc:491]: can you add some parentheses here to make this less ambiguous?

Comment (deanlee) [reply selfdrive/ui/qt/maps/map.cc]: added.this is also a bug if without parentheses

Comment (adeebshihadeh) [reply selfdrive/ui/qt/maps/map.cc]: is there a compiler flag we can set to make the build fail?

Comment (deanlee) [reply selfdrive/ui/qt/maps/map.cc]: Not sure, hope there is such a flag. i'll look up the documentation and add it if there is.

File: selfdrive/ui/qt/maps/map.cc [hunk 3]
```diff
@@ -407,12 +408,33 @@ MapInstructions::MapInstructions(QWidget * parent) : QWidget(parent) {
...
+
   QPalette pal = palette();
   pal.setColor(QPalette::Background, QColor(0, 0, 0, 150));
   setAutoFillBackground(true);
   setPalette(pal);
 }
 
+void MapInstructions::buildPixmapCache() {
+  QDir dir("../assets/navigation");
+  for (QString fn : dir.entryList({"*.png"}, QDir::Files)) {
```
Comment (adeebshihadeh) [inline selfdrive/ui/qt/maps/map.cc:421]: use ICON_SUFFIX instead of hardcoding .png

File: selfdrive/ui/qt/maps/map.cc [hunk 4]
```diff
@@ -516,7 +525,7 @@ void MapInstructions::updateInstructions(cereal::NavInstruction::Reader instruct
     if (!label->parentWidget()) {
       lane_layout->addWidget(label);
     }
-    label->setPixmap(loadPixmap(fn + ICON_SUFFIX, {125, 125}, Qt::IgnoreAspectRatio));
+    label->setPixmap(pixmap_cache[fn + ICON_SUFFIX]);
```
Comment (adeebshihadeh) [inline selfdrive/ui/qt/maps/map.cc:528]: remove ICON_SUFFIX from the key?

**Review by adeebshihadeh — CHANGES_REQUESTED** [review]

File: selfdrive/ui/qt/maps/map.cc [hunk 5]
```diff
@@ -411,6 +412,28 @@ MapInstructions::MapInstructions(QWidget *parent) : QWidget(parent) {
...
+    if (key.contains("turn_")) {
+      pixmap_cache[key] = pm.scaled({125, 125}, Qt::IgnoreAspectRatio, Qt::SmoothTransformation);
+    } else {
+      pm = pm.scaledToWidth(200, Qt::SmoothTransformation);
+      pixmap_cache[key] = pm;
+      // for rhd, reflect direction and then flip
+      if (key.contains("_left")) {
+        pixmap_cache["rhd_" + key.replace("_left", "_right")] = pm.transformed(QTransform().scale(-1, 1));
+      } else if (key.contains("_right")) {
+        pixmap_cache["rhd_" + key.replace("_right", "_left")] = pm.transformed(QTransform().scale(-1, 1));
```
Comment (sshane) [inline selfdrive/ui/qt/maps/map.cc:433]: would directions flip with rhd?

**Comment by incognitojam** [comment]
The scaling appears to be wrong for some instructions:

This PR:
image

master:
image

The icons seem to be correct (or at least, same as master in replay) for RHD. Will test in the car just to make sure. The mapbox directions don't seem perfect for roundabouts here.

**Comment by deanlee** [comment]
Fixed it, I didn't realize there were some icon files that were shared. It just scales differently when used.

**Comment by deanlee** [comment]
I didn't realize there were some icon files that were shared. they  just scales differently when used.

**Review by incognitojam — APPROVED** [review]

**Pull request merged**